### PR TITLE
REGR: ufunc with DataFrame input not passing all kwargs

### LINKED ISSUE #40662: BUG: numpy functions (eg, np.add) on DataFrames with 'out' parameter no longer work properly
- [x] I have checked that this issue has not already been reported.

- [x] I have confirmed this bug exists on the latest version of pandas.

- [ ] (optional) I have confirmed this bug exists on the master branch of pandas.

---

#### Code Sample, a copy-pastable example

```python
# works
foo = np.array([1,2,3,4]).reshape(2,2)
bar = np.zeros_like(foo)
np.add(foo, 1, out=bar)
print(bar)

# doesn't work. bar remains an array of zeros
foo = np.array([1,2,3,4]).reshape(2,2)
df = pd.DataFrame(foo)
bar = np.zeros_like(df)
np.add(df, 1, out=bar)
print(bar)
```

#### Problem description

Many numpy functions accept an `out` parameter as a way of storing the result. In the current pandas version (1.2.3), when the numpy function takes in a DataFrame as input, it no longer stores the result to the `out` array.

In pandas 1.0.1 (and earlier, on my other machine), the behavior works as expected. (see Expected Output). I am using numpy==1.19 for both machines.

Curiously, note that this problem is not seen for `pd.Series`. The below code works fine:

```python
foo = np.array([1,2,3,4])
series = pd.Series(foo)
bar = np.zeros_like(series)
np.add(series, 1, out=bar)
print(bar)
```

One workaround is to instead use `np.add(df.values, 1, out=bar)` but I am using several packages that require the older behavior.

#### Expected Output

```python
>>> print(bar)

[[2 3]
 [4 5]]
```

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit           : f2c8480af2f25efdbd803218b9d87980f416563e
python           : 3.7.4.final.0
python-bits      : 64
OS               : Linux
OS-release       : 3.10.0-1062.1.1.el7.x86_64
Version          : #1 SMP Fri Sep 13 22:55:44 UTC 2019
machine          : x86_64
processor        : x86_64
byteorder        : little
LC_ALL           : None
LANG             : en_GB.UTF-8
LOCALE           : en_GB.UTF-8

pandas           : 1.2.3
numpy            : 1.19.5
pytz             : 2019.3
dateutil         : 2.8.1
pip              : 19.3.1
setuptools       : 41.4.0
Cython           : 0.29.21
pytest           : None
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : 4.5.0
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : 2.10.3
IPython          : 7.11.1
pandas_datareader: None
bs4              : 4.9.1
bottleneck       : None
fsspec           : 0.8.2
fastparquet      : None
gcsfs            : None
matplotlib       : 3.1.3
numexpr          : 2.7.0
odfpy            : None
openpyxl         : 3.0.4
pandas_gbq       : None
pyarrow          : 0.13.0
pyxlsb           : None
s3fs             : 0.5.1
scipy            : 1.4.1
sqlalchemy       : 1.3.13
tables           : 3.6.1
tabulate         : 0.8.7
xarray           : None
xlrd             : 1.2.0
xlwt             : None
numba            : 0.49.1

</details>

### PR DESCRIPTION
**Pull request by mzeitlin11**
- [x] closes #40662
- [x] tests added / passed
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#code-standards) for how to run them
- [x] whatsnew entry

Credit to @attack68 for the fix, just figured we should try to get this is in for 1.2.4 if possible since we know the fix.

### CODE REVIEW

File: pandas/tests/frame/test_ufunc.py
Comment (phofl) [inline pandas/tests/frame/test_ufunc.py:63]: Does it make sense to parametrize over the functions?

**Review by phofl — COMMENTED** [review]
minor comment, otherwise lgtm

File: pandas/tests/frame/test_ufunc.py
Comment (mzeitlin11) [reply pandas/tests/frame/test_ufunc.py]: Good call, have parameterized

File: pandas/tests/frame/test_ufunc.py
Comment (simonjayhawkins) [inline pandas/tests/frame/test_ufunc.py:81]: can you also test the return type/values of the operation itself.

**Review by simonjayhawkins — CHANGES_REQUESTED** [review]
Thanks @mzeitlin11 for the PR. we want to make sure that the return type is unchanged.

File: pandas/tests/frame/test_ufunc.py
Comment (mzeitlin11) [reply pandas/tests/frame/test_ufunc.py]: done

File: doc/source/whatsnew/v1.2.4.rst
Comment (simonjayhawkins) [inline doc/source/whatsnew/v1.2.4.rst:24]:
```suggestion
- Fixed regression in NumPy ufuncs such as ``np.add`` not passing through all arguments for 2-dimensional input (:issue:`40662`)
```

and maybe replace 2-dimensional input with :class:`DataFrame`?

**Review by simonjayhawkins — APPROVED** [review]
Thanks @mzeitlin11 lgtm

File: doc/source/whatsnew/v1.2.4.rst
Comment (mzeitlin11) [reply doc/source/whatsnew/v1.2.4.rst]: Yep will do

File: doc/source/whatsnew/v1.2.4.rst
Comment (mzeitlin11) [inline doc/source/whatsnew/v1.2.4.rst:24]:
```suggestion
- Fixed regression in NumPy ufuncs such as ``np.add`` not passing through all arguments for :class:`DataFrame` (:issue:`40662`)
```

File: pandas/tests/frame/test_ufunc.py
Comment (phofl) [reply pandas/tests/frame/test_ufunc.py]: subtract maybe too?

Comment (mzeitlin11) [reply pandas/tests/frame/test_ufunc.py]: Sure, have added that

**Review by phofl — APPROVED** [review]

**Comment by phofl** [comment]
lgtm Ci failure is because of numpy dev issue

File: pandas/core/arraylike.py
Comment (simonjayhawkins) [inline pandas/core/arraylike.py:360]: the else clause is reached for a DataFrame when not `(len(inputs) > 1 or ufunc.nout > 1)`.

The `result = mgr.apply(getattr(ufunc, method))` there for `__call__` also fails to pass along **kwargs

Is it straightforward to add to the paramterised here to exercise that path.

Comment (mzeitlin11) [reply pandas/core/arraylike.py:360]: Yep, I think so...will let you know if I run into any issues

Comment (mzeitlin11) [reply pandas/core/arraylike.py:360]: Added a test, but had to xfail because `out` is not written to correctly. The written result becomes transposed with the block-wise apply, I'm guessing because of the following relationship:
```
arr = np.array([[1, 2], [3, 4]])
df = pd.DataFrame(arr)
print(df.values)
print(df._mgr.blocks[0].values)
```
gives 
```
[[1 2]
 [3 4]]
[[1 3]
 [2 4]]
```

File: pandas/tests/frame/test_ufunc.py
Comment (jreback) [inline pandas/tests/frame/test_ufunc.py:99]: do we have any documentation that out is actually a ndarray? this is a very strange result. At the very least document this, and let's open an issue. This should work with out=DataFrame, or simply raise (preferred)

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tests/frame/test_ufunc.py
Comment (simonjayhawkins) [reply pandas/tests/frame/test_ufunc.py:99]: yep. I thought this was strange. https://github.com/pandas-dev/pandas/issues/40662#issuecomment-817307177

should we defer to 1.2.5 to allow more discussion? or put 1.2.4 release on hold for a day or two?

Comment (jreback) [reply pandas/tests/frame/test_ufunc.py:99]: yeah let's defer this. i dont' think this is the correct behavior and we should actually fix it (which may mean that we simply do this for 1.3)

Comment (jreback) [reply pandas/tests/frame/test_ufunc.py:99]: moving to 1.2.5 (if we do it)

File: pandas/core/arraylike.py
Comment (jorisvandenbossche) [reply pandas/core/arraylike.py:360]: Yeah, basically when `out` is specified, I think we should simply not call `mgr.apply`, see https://github.com/pandas-dev/pandas/issues/39275, https://github.com/pandas-dev/pandas/pull/39260/files for a recent similar case where certain additional keyword arguments cannot be handled on a block-by-block case.

File: pandas/core/arraylike.py
Comment (jorisvandenbossche) [inline pandas/core/arraylike.py:367]:
```suggestion
        if method == "__call__" and kwargs.get("out", None) is None:
```

(untested, but if

Comment (jorisvandenbossche) [reply pandas/core/arraylike.py:367]: Or maybe more simply check for any `kwargs`, so only do this if `kwargs` is empty (so `and not kwargs`)

File: pandas/tests/frame/test_ufunc.py
Comment (jorisvandenbossche) [reply pandas/tests/frame/test_ufunc.py:99]: > This should work with out=DataFrame, or simply raise (preferred)

This does raise when passing a DataFrame to `out` (as numpy expects an array as `out` argument), but what is being tested here is the return value of the ufunc *in case of* passing an array to `out`. In that case, `out` is also being returned, and since `out` is an ndarray, the return value is also an ndarray. 

(to me this seems correct behaviour, and thus I don't think this needs to hold up merging this for the release)

Comment (simonjayhawkins) [reply pandas/tests/frame/test_ufunc.py:99]: the return type is a DataFrame when out is specified and this has been added to the test

```
 expected = pd.DataFrame(expected)
 tm.assert_frame_equal(result, expected)
```

Comment (jreback) [reply pandas/tests/frame/test_ufunc.py:99]: the very fact that you can specify out as a ndarray while the input is a DataFrame is just very strange behavior and should simply not be allowed. This needs fixing and much more testing.

File: pandas/core/arraylike.py
Comment (jorisvandenbossche) [reply pandas/core/arraylike.py]: I quickly pushed this edit myself, as it fixes the xfail case in the tests you added

File: pandas/tests/frame/test_ufunc.py
Comment (jorisvandenbossche) [reply pandas/tests/frame/test_ufunc.py:95]: > the return type is a DataFrame when out is specified and this has been added to the test

Whoops, sorry, was testing in the wrong env. Indeed, the return value is properly wrapped in a DataFrame. Indeed, now it correctly returns a DataFrame while the `out` ndarray *is* updated

> the very fact that you can specify out as a ndarray while the input is a DataFrame is just very strange behavior and should simply not be allowed. This needs fixing and much more testing.

Before DataFrame implemented this itself (with `__array_ufunc__`), any input to a numpy ufunc was coerced to an array, and `out` could also be an ndarray. This is longstanding behaviour (and IMO also fine that this works). 

If we want to change this, it can be deprecated, but for now fixing this severe regression (you silently get different values) seems the best thing to do. 
(the fix is also tiny, it's just passing through the `kwargs`, which was an oversight that this was not done before)

And note this is not only about `out`, but also about other kwargs being ignored.

**Review by jorisvandenbossche — APPROVED** [review]

File: pandas/tests/frame/test_ufunc.py
Comment (simonjayhawkins) [reply pandas/tests/frame/test_ufunc.py:95]: > And note this is not only about `out`, but also about other kwargs being ignored.

that was the response to my comment in #40662 https://github.com/pandas-dev/pandas/issues/40662#issuecomment-817310283 so was happy to approve and I am happy to include in 1.2.4 (but I still also think the out behavior is strange)

Comment (jreback) [reply pandas/tests/frame/test_ufunc.py:95]: ok i guess ok for 1.2.4 as this fixes the regression. but this is REALLY strange and we need to fix this (e.g. out being an ndarray), this is just plain wrong, yes this is a numpy ufunc but is so unexpected, we can simply raise an error if out is not a DataFrame I think would make this way better.

Comment (jreback) [reply pandas/tests/frame/test_ufunc.py:95]: on 2nd thought, it think this should be fixed before releasing.

Comment (jreback) [reply pandas/tests/frame/test_ufunc.py:95]: so again let's not do a last minute PR , not reason.

Comment (jorisvandenbossche) [reply pandas/tests/frame/test_ufunc.py:95]: Jeff, this is a tiny bugfix (look at the diff, it's just passing through `**kwargs`), that fixes a quite serious regression (you silently get wrong values when passing any additional keyword argument). 
The behavior you are objecting to is a long-standing behaviour, that we can discuss separately, but is not specific to this PR (this PR is *not* only for `out`, but fixes the regression for any keyword argument, such as `where`)

Comment (jorisvandenbossche) [reply pandas/tests/frame/test_ufunc.py:95]: > on 2nd thought, it think this should be fixed before releasing.

What needs to be fixed? 
If you want to change how we handle the `out` keyword (eg raising an error if it isn't a DataFrame), that can go through a deprecation cycle and thus is for another PR targeting 1.3 (and if so, please open an issue about it)

Comment (jreback) [reply pandas/tests/frame/test_ufunc.py:95]: I disagree. This is PR enables this itself by passing thru the out keyword. You say its a bug fix, but this a problem. Releasing this PR is enabling this terrible behavior.

Comment (jorisvandenbossche) [reply pandas/tests/frame/test_ufunc.py:95]: This "terrible" behaviour is *long-standing* (and IMO correct) behaviour on which users rely. Again, if we want to change this, we can discuss it and it can go through a deprecation cycle. But for now let's fix the regression.

Comment (simonjayhawkins) [reply pandas/tests/frame/test_ufunc.py:95]: and would make DataFrame consistent with Series.. on master..

```
>>> foo = np.array([1, 2, 3, 4])
>>> foo
array([1, 2, 3, 4])
>>>
>>> ser = pd.Series(foo)
>>> ser
0    1
1    2
2    3
3    4
dtype: int32
>>>
>>> bar = np.zeros_like(ser)
>>> bar
array([0, 0, 0, 0])
>>>
>>> np.add(ser, 1, out=bar)
0    2
1    3
2    4
3    5
dtype: int32
>>>
>>> bar
array([2, 3, 4, 5])
>>>
```

Comment (jorisvandenbossche) [reply pandas/tests/frame/test_ufunc.py:95]: @jreback do you have time to hop on a quick call to resolve this dispute?

Comment (jreback) [reply pandas/tests/frame/test_ufunc.py:95]: > This "terrible" behaviour is long-standing (and IMO correct) behaviour on which users rely. 

how is this correct in any way? this was a bug from the original impl.

Comment (jorisvandenbossche) [reply pandas/tests/frame/test_ufunc.py:95]: That was not a bug, that's how numpy functions work: they coerce array-like input to arrays. So whether one of the arguments was a DataFrame vs an ndarray, did not have any effect on allowing an `out` argument or not

Comment (jreback) [reply pandas/tests/frame/test_ufunc.py:95]: ok fine. how about a follow to make this *really* clear (in docs) on master. I would however also deprecate / remove this as it not intuitive at all (e.g. we do not have an out argument anywhere else)

**Review by jreback — APPROVED** [review]

**Comment by jorisvandenbossche** [comment]
Thanks @mzeitlin11 !

File: pandas/tests/frame/test_ufunc.py
Comment (jorisvandenbossche) [reply pandas/tests/frame/test_ufunc.py:95]: > e.g. we do not have an out argument anywhere else

This keyword is not in a pandas function or method, but in a NumPy function (which has `out` in many places)

**Comment by simonjayhawkins** [comment]
@meeseeksdev backport 1.2.x

File: pandas/tests/frame/test_ufunc.py
Comment (jreback) [reply pandas/tests/frame/test_ufunc.py:95]: i know, but this is *very* confusing if someone is doing this (and doesnt; realize it).

File: pandas/core/arraylike.py
Comment (mzeitlin11) [reply pandas/core/arraylike.py]: Thanks for adding this!

**Pull request merged**
